qualtran/bloqs/multiplexers/apply_lth_bloq.py

```diff
@@ -0,0 +1,100 @@
+#  Copyright 2024 Google LLC
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      https://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+from functools import cached_property
+from typing import Sequence, Tuple
+
+import cirq
+import numpy as np
+from attrs import field, frozen, validators
+
+from qualtran import Bloq, bloq_example, BloqDocSpec, BoundedQUInt, QBit, Register, Side
+from qualtran.bloqs.multiplexers.unary_iteration_bloq import UnaryIterationGate
+
+
+@frozen
+class ApplyLthBloq(UnaryIterationGate):
+    r"""A SELECT operation that executes one of a list of bloqs $U_l$ based on a quantum index:
+
+    $$
+    \mathrm{SELECT} = \sum_{l}|l \rangle \langle l| \otimes U_l
+    $$
+
+    This bloq uses the unary iteration scheme to apply `ops[selection]` controlled on an optional single-bit `control` register.
+
+    Args:
+        ops: List of bloqs to select from, each with identical registers that are all THRU.
+        control: If True, a singly controlled gate is constructed.
+
+    Registers:
+        selection: The index of the bloq in `ops` to execute.
+        control: The control bit if specified above.
+        [user_spec]: The output registers of the bloqs in `ops`.
+
+    References:
+        [Encoding Electronic Spectra in Quantum Circuits with Linear T Complexity](https://arxiv.org/abs/1805.03662).
+        Babbush et. al. (2018). Section III.A. and Figure 7.
+    """
+
+    ops: Tuple[Bloq, ...] = field(
+        converter=lambda x: x if isinstance(x, tuple) else tuple(x), validator=validators.min_len(1)
+    )
+    control: bool = False
+
+    def __attrs_post_init__(self):
+        if not all(u.signature == self.ops[0].signature for u in self.ops):
+            raise ValueError("All ops must have the same signature.")
+        if not all(r.side == Side.THRU for u in self.ops for r in u.signature):
+            raise ValueError("All ops must have only THRU registers.")
```

Comment:
What's the motivation of this bloq with these restrictive constraints? Where will this be used?

Reply:
This bloq will be used to implement linear combinations of block encodings given a list of those constituent block encodings and their corresponding coefficients. 

Before creating this bloq, I checked the existing implementations of Select oracles, and did not find any whose interface simply accepts a list of bloqs and coherently indexes into that list as needed for the block encoding.

I could be wrong, but the constraints here (all ops in the list have the same signature of only THRU registers) are strictly weaker than certain other existing Select oracles (e.g. `ApplyGateToLthQubit`, which additionally requires the gates to be single-qubit). 

Is there a way in which you find these constraints to be inappropriately restrictive?

Reply:
As you mentioned in your main comment:

> Wouldn't the restriction that all block encodings in the LCU should have the same signature be restrictive?

Could you give an example of this condition being overly restrictive? Our quantum interface for block encodings already requires these signatures to be identical up to the sizes of `ancilla` and `resource` registers. In this case, the implementation of the `LinearCombination` block encoding is responsible for harmonizing these registers --- I've implemented that logic as part of a later PR.

The reason I didn't generalize the logic to harmonize registers and put it into this bloq is because the logic seems to be case-dependent. For example, the bits of the `ancilla` register are reused among the constituent block encodings, but it's unclear whether this should hold of arbitrary other registers in the signatures of bloqs that `ApplyLthBloq` selects from.

Reply:
Suppose you want to construct a block encoding for an SK model hamiltonian. The hamiltonian is represented as

$$
H = \sum_{i < j} w_{i, j} Z_{i} Z_{j}
$$

We can rewrite the hamiltonian as a summation over a single contiguous index $k$ that goes from $0$ to $\frac{n(n-1)}{2}$ s.t.

$$
H = \sum_{k} w_{k} Z_{k, 0} Z_{k, 1}
$$

Now I have an LCU over $\frac{n(n-1)}{2}$ terms. Here each term is a `Unitary` block encoding with `U=ZZ` gate; i.e. `Unitary(U=ZZ, alpha=1, ancilla_bitsize=0, resource_bitsize=0, epsilon=0)`. The `system_bitsize` would be 2 since the unitary acts on 2 qubits. 

Now, if I do the more "intuitive" thing and simply construct an array of size $k$ where each element is a unitary block encoding as described above and pass it to your `LCUBlockEncoding` bloq to construct a block encoding for my hamiltonian, then I'll get a wrong result. This is because the current interface does not encode information about "which qubits you want to apply the i'th bloq on". To encode this information, I'll have to wrap all these unitaries in a wrapper which encodes in the information about `Z_{k, 0} Z_{k, 1}` such that each bloq acts on all the $k$ qubits instead of only two qubits and then pass this list of wrappers to the `LCUBlockEncoding` interface. 


Instead of forcing users to wrap every term of the LCU in a wrapper that acts on the entire system register; we can also provide an abstraction which accepts a `Bloq` and maybe an ndarray of indices that tell you on which subset of qubits from your system register you want to apply the bloq on to construct the block encoding. 

But I'm just curious to understand better if you've already thought about this problem and have a preferred approach one way or the other. 

> Before creating this bloq, I checked the existing implementations of Select oracles, and did not find any whose interface simply accepts a list of bloqs and coherently indexes into that list as needed for the block encoding.

This is because all existing Select oracles are for specific problem instances where we have more problem specific information about "where to apply the i'th bloq" and so we didn't need a generalized version of indexing into the i'th bloq given a list of bloqs.

Reply:
> Instead of forcing users to wrap every term of the LCU in a wrapper that acts on the entire system register; we can also provide an abstraction which accepts a Bloq and maybe an ndarray of indices that tell you on which subset of qubits from your system register you want to apply the bloq on to construct the block encoding.

Thanks for giving the example. It makes sense and was not a case that I had previously considered.

If it's OK with you, I'd like to add support for such unitaries that act over distinct parts of the system register in a follow-up PR after this one.

Reply:
Thanks for the write-up @tanujkhattar . If I understand it correctly, you're objecting to wrapping each operation so it has a consistent set of registers. This introduces a level of indirection; but makes a very conceptually simple bloq composition. If I understand it correctly, it would additionally unblock having a symbolic decomposition of a SELECT oracle which has myriad benefits

Reply:
> If I understand it correctly, you're objecting to wrapping each operation so it has a consistent set of registers

Quoting from my comment above "But I'm just curious to understand better if you've already thought about this problem and have a preferred approach one way or the other." 

I am not objecting to anything. I was just pointing out that we will have cases where the LCU block encoding would be composed of parts which act on different parts of the system. We should either provide a wrapper or an abstraction to capture which bloq acts on which part of the subsystem. Both of these would be fine ways of doing things. 

If we use a wrapper, we also risk exposing an API that users can use incorrectly -- see my example above on following the more intuitive approach and how that leads to a bloq which doesn't raise an error on construction but is an incorrect bloq encoding.